### ISSUE TITLE
Question: MQTT Client Ids in authentication

### ISSUE THREAD

**Issue by weeco** [erlio/vernemq#825]
I noticed that VerneMQ requires a static client ID bound to a login:

```sql
INSERT INTO vmq_auth_acl 
    (mountpoint, client_id, username, 
     password, publish_acl, subscribe_acl)
VALUES 
    ('', 'test-client', 'test-user', PASSWORD('123'), NULL, NULL);
```

If I use `test-client` as clientId the authentication works just fine. However my current understanding of the MQTT protocol is that the clientId needs to be unique for every client connection to the broker as the identifier is used to close half-open TCP connections and session management. I intend to run multiple clients using authentication.

I believe I am looking for an approach how I could accept all clientIds for a given username/password combination. Is this possible with VerneMQ?

**Comment by larshesel** [comment]
Hi @weeco 

The current sql plugins require the client_id to be fixed. The reason for this is that only username and password is authenticated, then it's possible for a malicous client to steal sessions from other clients if it is possible to guess the client id of those other clients. So this is something to be aware about when relaxing this requirement.

The sql plugins doesn't support relaxing this requirement, but the logic is implemented in lua and you can easily change them if you want to: https://github.com/erlio/vernemq/blob/master/apps/vmq_diversity/priv/auth/postgres.lua#L70-L82

But as mentioned above it's important to makes sure each client has it's own client_id. So one solution could be to fix the client id to be the same as the username and populate the database with that.

I hope this helps.

**Comment by weeco** [comment]
Thanks for your quick answer @larshesel 
I don't understand your reasoning why this might be a security issue. Are you saying it's an issue because a single user (of potentially 100 further users) could use ANY clientId (even clientIds which are in use by one of the other 100 users)?

I am surprised that this isn't a common scenario where you want to protect your MQTT broker with username/password but you want to run multiple authorized clients. Even though I am not the developer of the MQTT services, as far as I got the client libraries usually create a random uuid for each initiated connection.

> But as mentioned above it's important to makes sure each client has it's own client_id. So one solution could be to fix the client id to be the same as the username and populate the database with that.

My understanding is that this wouldn't help because multiple clients would again use the same clientId and kick each other, no?

**Comment by larshesel** [comment]
The MQTT spec says that the sole identifier for the session is the client id. And if a client connects for which there is an existing session then the old clients gets disconnected if it's still connected and takes over any left over state (subscriptions etc) depending on how the `clean_session` flag was set.

So if you have two clients that are authenticated only with username and password but are otherwise free to pick their client id, then they can interfere with each other if they can guess  the client id of the other. If this happens then the least worst case is just that the other client gets disconnected. The worst case is that the other client takes over the the subscriptions which the first client made if `clean_session=false` was used.

**Comment by weeco** [comment]
I think I understood this, however the chance of a uuid collision is really low when the client uses proper UUIDs as client id. No matter what I need to figure out a way how one could connect with as many clients as desired using authentication.

Since VerneMQ forces me to provide a client id along with username/password I can not allow multiple simultaneous connections for one username/password combination without modifying the lua script right?

**Comment by larshesel** [comment]
> Since VerneMQ forces me to provide a client id along with username/password I can not allow multiple simultaneous connections for one username/password combination without modifying the lua script right?

That is correct.

Note, that with the current replication mechanism (which is also enabled if running a single node) client-ids are never forgotten, so using random client ids will lead to increased resource usage over time - this limitation will be lifted with the new replication mechanism we're developing for VerneMQ 2.0 (which may be back-ported to the 1.x series).

**Comment by tobdub** [comment]
What about some kind of wildcard for the client id so that one user can have multiple clients?

**Comment by bpair** [comment]
I realize this issue is closed but the way VereMQ implements this seems to be opposite of how other projects are implementing. I believe the key quote from above is "The MQTT spec says that the sole identifier for the session is the client id.", and VerneMQ is being very strict about this. Certainly it seems plausible to combine the clientId with the username (or some propietary hash of the username/password) to establish a unique session. Otherwise it sounds like a client that authenticates with username=John@companyA.com and chooses a clientId=John would have their session stolen when a client with userame=John@companyB.com uses a clientId=John. That is indeed a problematic situation!

**Comment by larshesel** [comment]
Hi @bpair - I'm not sure I follow completely - what do you propose?

**Comment by bpair** [comment]
I was suggesting that the clientId not be the sole session identifier. Instead a combination of clientId and username should identify the session with ACL based on the username and not the clientId. It has been my experience that the MQTT standard is to definitely kick out active sessions when a new login for the same clientId begins. However, different usernames can use the same clientId without any issue, and dynamic clientIds can be used to create several "threads" with the same username.

**Comment by larshesel** [comment]
In the diversity plugins we use the combination of client-id, username and password which must all be correct in order for the client to connect. So that's an example where we're using all the information to authenticate and authorize. But the client-id *is* the only/unique identifier for the session - I haven't found any evidence in the spec to the contrary. I'd love a pointer to it, though.

**Comment by bpair** [comment]
But what I think you are saying is that two different clients, with different usernames and passwords, can affect each other if they use the same clientId. I just don't think that is the desired situation defined by the MQTT spec.

**Comment by bpair** [comment]
I scanned over the MQTT5 spec, and I think we are both partially correct. The section on ClientId, 3.1.3.1, does state that the ClientID is supplied by the client and that each client connecting to the server has a unique client id.

> **3.1.3.1 Client Identifier (ClientID)**

> The Client Identifier (ClientID) identifies the Client to the Server. Each Client connecting to the Server has a unique ClientID. The ClientID MUST be used by Clients and by Servers to identify state that they hold relating to this MQTT Session between the Client and the Server [MQTT-3.1.3-2].

However there is a section about letting a client supply no clientId and also a dynamic clientId:

> A Client implementation could provide a convenience method to generate a random ClientID.
Clients using this method should take care to avoid creating long-lived orphaned Sessions.

Those sections, suggest to me, that using static clientIds in a database for ACL is not the intended implementation. In the Authentication section the following is stated:

> Authorization may be based on information provided by the Client such as User Name, the hostname/IP address of the Client, or the outcome of authentication mechanisms.

> In particular, the implementation should check that the Client is authorized to use the Client Identifier as this gives access to the MQTT Session State (described in section 4.1). This authorization check is to protect against the case where one Client, accidentally or maliciously, provides a Client Identifier that is already being used by some other Client. 

**So the server should not allow a connection from a different username to use the same clientId.**
In short I believe you are correct that the clientId is used as the session identifier and must be unique, however the intention seems to be to allow dynamically generated clientIDs (either by the client or the server) and therefore not to use static clientIDs for ACL. In fact it appears to be the decision of the client to reuse a clientId or not. Having the clientId in in a database for ACL seems incorrect.

My main reason for bringing this up is a client running in kubernetes that may scale up or down. Having static clientIds make this scenario much more difficult.

**Comment by bpair** [comment]
We were thinking about adding records to the vmq_acl_auth table prior to each login (e.g. when the kubernetes cluster starts) but we noticed the following open issue:
> Note, that with the current replication mechanism (which is also enabled if running a single node) client-ids are never forgotten, so using random client ids will lead to increased resource usage over time - this limitation will be lifted with the new replication mechanism we're developing for VerneMQ 2.0 (which may be back-ported to the 1.x series)."  https://github.com/vernemq/vernemq/issues/825

**Comment by larshesel** [comment]
@bpair use the vmq_swc metadata backend instead: https://vernemq.com/blog/2019/02/01/vernemq-next-level-replication.html

> In fact it appears to be the decision of the client to reuse a clientId or not. Having the clientId in in a database for ACL seems incorrect.

I understand it might be more difficult, but it is the safe thing to do as depending on the scheme the client-ids are chosen they might be easily guessable. If a device is hacked and become malicious it can then freely replace the client-id and can take over and steal session from under the feet of other clients.

If you need a different behaviour it is fairly easy to customize the lua scripts to suit your needs.

**Comment by bpair** [comment]
I really appreciate your willingness to discuss this.

I do feel you are not implementing the MQTT5 spec if you allow clients with different usernames to use the same clientId for an active session or persistent session. I do not think this follows the MQTT5 spec.

> In particular, the implementation should check that the Client is authorized to use the Client Identifier as this gives access to the MQTT Session State (described in section 4.1). This authorization check is to protect against the case where one Client, accidentally or maliciously, provides a Client Identifier that is already being used by some other Client.

Our clients always use clean sessions and mostly QoS0 messages. We do use QoS1 as well, but we do not expect a client that is disconnected to receive any QoS1 messages when reconnecting. We absolutely cannot have a situation where client A sees messages for client B because of a clientId clash. Is there an approach that absolutely prevents this? Maybe we should use all QoS0, but we are optimistic that our clients will have stable long lasting connections. 

If client A can never see messages for client B then we would like for these clients to be able to dynamically generate as many clientIds as they need without negatively impacting the system with old abandoned sessions or old abandoned ACL entries.

**Comment by larshesel** [comment]
I think we're talking past each other. The `vmq_diversity` database plugin authenticates using username, password and client-id. If you decide to add a row with different usernames and passwords but the same client-id then that is up to you.

If you're using dynamic client-ids (generated by the broker) - then you probably need a custom auth plugin to handle that - `vmq_diversity` wasn't built for that use-case.

And with that said - I think the slack channel would be a better place for a discussion like this ;-)

**Comment by bpair** [comment]
Fair enough. We have added a unique constraint on the client ID in the mongo vmq_acl_auth collection.

**Comment by pankaj-dahiya** [comment]
We are using MQTT for IOT purpose, obviously! We have more then 6-7 millions of devices, sending data to MQTT using dynamic client-ids (generated by the broker). As of now, we are using auth using files and thinking to move towards auth using db but seems like this can't be an option as client-ids are to be static when using db!

**Comment by epicvinny** [comment]
Hi @bpair ! Same scenario here, trying to run multiple authenticated subscribers on Kubernetes using HPA (through rancher). 

My solution was: my subscriber app deployment has a job to create ACL on postgres based on pod name - it runs everytime a new pod is created, so it consume an environment variable with the unique pod name and use it to create a new user as well use it inside the subscriber app as clientid.

**Comment by ryanshow** [comment]
We also ran into this issue, were we have credentials set up for an in-browser websocket client that may have an arbitrary number of instances. I ended up solving the problem with inspiration from the comments by @larshesel by modifying the postgres_cockroach_commons.lua script and changing the queries to support arbitrary text after the client_id.

Change this:
`client_id=$2`

To this:
`$2 LIKE client_id || '%'`

Note there are two queries you may have to modify, one for client side password hashing and one for server side. This works as long as you ensure that your client_ids are unique and are not subsets of each other. For example, you can not have a client_id of 'foo' and another client_id of 'foobar' (because clients connected with id 'foobar123' could potentially match both.)

In my case, I'm running vernemq in a docker instance, so I just mounted the `/vernemq/share` directory as a volume on the host and edited the lua files there.

I realize this solution may not be suitable for everyone, but it solved my use case nicely.

**Comment by tech-scorpio** [comment]
Hello, your solution seems to be very usefull. However, can you explain how you persist the changement in the docker container when the container restart?
I try to modify the script like you, but it seems that the script is loaded at the startup on vernemq. (can you confirm that?)  and when I restart the container the changement is lost.
I use an helm chart to install vernemq (https://github.com/vernemq/docker-vernemq/tree/master/helm/vernemq), it is maybe a different use case than yours but I thing your solution is the thing I need.

**Comment by ryanshow** [comment]
It's been a while since I posted this comment, but I think my solution was just to specify `/vernemq/share` as a mount so that the contents are persisted on the host filesystem between restarts. However, I'm unfamiliar with how helm works.. my use case uses docker-compose.

**Comment by maruno** [comment]
I just hit this limitation making it impossible for me to switch from EMQx to VerneMQ? This is a weird limitation to have, especially since both you and EMQx have the possibility to use the client-id as a variable in part of the ACL JSON... But then using the database it makes it a static?

In the context of IOT this to me makes little sense. I got devices that all share the same login credentials and are sharing the same security context. Each device has a unique client-ID to have their own session and one specific topic per device. To me this seems like a perfectly valid and common use case in MQTT? I now provision a device and it automatically gets added into all my systems, but now I wouldn't be able to with VerneMQ. I would need to extract the microcontroller device-ID and insert a copy of the vmq_auth_acl rule for every client-id...

I completely understand security considerations and in some context this could be valid, but you are not really leaving an option? Why is it not simply possible to constrain to a client-id as an option, for example by making the column nullable? Personally if I expected a unique per device ACL I'd make a different user for each, what else is the point of being able to share a user?

**Comment by ioolkos** [comment]
@maruno 

1. Regarding "weird" limitation: it is a recommendation of the MQTT spec to enforce alignement of ClientId, Username, Password
2. The Lua scripts are meant to be edited. You can take out the ClientId check there easily.
Note that this will allow any device having a Username and a Password to pose as any ClientId.

Also note the default rule thing in this PR: https://github.com/vernemq/vernemq/commit/4476607adf9c6f8ddde9a82310810db690c38ab7

Will review PRs with improvements or better ideas anytime.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by maruno** [comment]
@ioolkos The linked PR seems just what I needed to allow any client-id! :) I'll try it out, still figuring out some of the configuration! I'll see if I can make a PR at least to vmq-docs with some of the things I encountered when I finished replacing my old broker

**Comment by seetherisingazeez** [comment]
@ioolkos @maruno  I replaced the /share/lua/auth/postgres_cockroach_commons.lua file with the new file available in the linked PR above. I'm using docker. This doesn't seem to work though. Am I missing something here?

**Comment by maruno** [comment]
> @ioolkos @maruno  I replaced the /share/lua/auth/postgres_cockroach_commons.lua file with the new file available in the linked PR above. I'm using docker. This doesn't seem to work though. Am I missing something here?
> 

I don't remember I actually had to compile a different version changing any code? It worked fine here as long as the client id in the database row is set to the string '*'.
